runtime/runtime_test.go

```diff
@@ -234,3 +242,129 @@ func testRuntimeProcessWatchEventPolicyError(t *testing.T, asBundle bool) {
+	exp := &report.DataResponse{Latest: report.ReleaseDetails{
+		Download:     "https://openpolicyagent.org/downloads/v100.0.0/opa_darwin_amd64",
+		ReleaseNotes: "https://github.com/open-policy-agent/opa/releases/tag/v100.0.0",
+	}}
+
+	// test server
+	baseURL, teardown := getTestServer(exp, http.StatusOK)
+	defer teardown()
+
+	version.Version = "v0.20.0"
+	expected := "\n# A newer version of OPA is available for download !\n" +
+		"# OPA version v100.0.0 is now available. Current version v0.20.0\n" + "\n" +
+		"# Download OPA: https://openpolicyagent.org/downloads/v100.0.0/opa_darwin_amd64\n" +
+		"# Release Notes: https://github.com/open-policy-agent/opa/releases/tag/v100.0.0\n\n> "
+
+	testCheckOPAUpdate(t, baseURL, expected)
+}
+
+func TestCheckOPAUpdateLoopBadURL(t *testing.T) {
+	testCheckOPAUpdateLoop(t, "http://foo:8112", "Unable to send OPA version report")
+}
+
+func TestCheckOPAUpdateLoopNoUpdate(t *testing.T) {
+	// test server
+	baseURL, teardown := getTestServer(nil, http.StatusOK)
+	defer teardown()
+
+	testCheckOPAUpdateLoop(t, baseURL, "OPA version is up-to-date !")
+}
+
+func TestCheckOPAUpdateLoopWithNewUpdate(t *testing.T) {
+	exp := &report.DataResponse{Latest: report.ReleaseDetails{
+		Download:     "https://openpolicyagent.org/downloads/v100.0.0/opa_darwin_amd64",
+		ReleaseNotes: "https://github.com/open-policy-agent/opa/releases/tag/v100.0.0",
+	}}
+
+	// test server
+	baseURL, teardown := getTestServer(exp, http.StatusOK)
+	defer teardown()
+
+	version.Version = "v0.20.0"
+	testCheckOPAUpdateLoop(t, baseURL, "A newer version of OPA is available for download ! Current version v0.20.0")
+}
+
+func getTestServer(update interface{}, statusCode int) (baseURL string, teardownFn func()) {
+	mux := http.NewServeMux()
+	ts := httptest.NewServer(mux)
+
+	mux.HandleFunc("/v1/version", func(w http.ResponseWriter, req *http.Request) {
+		w.WriteHeader(statusCode)
+		bs, _ := json.Marshal(update)
+		w.Header().Set("Content-Type", "application/json")
+		w.Write(bs)
+	})
+	return ts.URL, ts.Close
+}
+
+func testCheckOPAUpdate(t *testing.T, url, expected string) {
+	t.Helper()
+	os.Setenv("OPA_TELEMETRY_SERVICE_URL", url)
+
+	ctx := context.Background()
+	rt := getTestRuntime(ctx, t)
+	var stdout bytes.Buffer
+	rt.Params.Output = &stdout
+
+	done := make(chan struct{})
+	go func() {
+		rt.checkOPAUpdate(ctx, done)
+		done <- struct{}{}
+	}()
+	<-done
+
+	if stdout.String() != expected {
+		t.Fatalf("Expected output:\"%v\" but got: \"%v\"", expected, stdout.String())
+	}
+}
+
+func testCheckOPAUpdateLoop(t *testing.T, url, expected string) {
+	t.Helper()
+	os.Setenv("OPA_TELEMETRY_SERVICE_URL", url)
+
+	ctx := context.Background()
+	rt := getTestRuntime(ctx, t)
+	var stdout bytes.Buffer
+	rt.Params.Output = &stdout
+
+	logrus.SetOutput(rt.Params.Output)
+	logrus.SetLevel(logrus.DebugLevel)
+
+	done := make(chan struct{})
+	go func() {
+		rt.checkOPAUpdateLoop(ctx, 1, done)
+	}()
+	time.Sleep(2 * time.Second)
```

Comment:
If we change the signature of `checkOPAUpdateLoop` to take in a duration, rather than the number of seconds, we could cut the test time down quite a bit.

Reply:
Done